io-engine/src/core/thread.rs

```diff
@@ -1,281 +1 @@
-/// the submitted work to completion.
-pub struct Mthread(NonNull<spdk_thread>);
-unsafe impl Send for Mthread {}
-impl From<*mut spdk_thread> for Mthread {
-    fn from(t: *mut spdk_thread) -> Self {
-        let t = NonNull::new(t).expect("thread may not be NULL");
-        Mthread(t)
-    }
-}
-
-impl Mthread {
-    pub fn get_init() -> Mthread {
-        Mthread(
-            NonNull::new(unsafe { spdk_thread_get_by_id(1) })
-                .expect("No init thread allocated"),
-        )
-    }
-
-    ///
-    /// With the given thread as context, execute the closure on that thread.
-    ///
-    /// Any function can be executed here however, this should typically be used
-    /// to execute functions that reference any FFI to SPDK.
-
-    pub fn new(name: String, core: u32) -> Option<Self> {
-        let name = CString::new(name).unwrap();
-
-        NonNull::new(unsafe {
-            let mut mask = CpuMask::new();
-            mask.set_cpu(core, true);
-            spdk_thread_create(name.as_ptr(), mask.as_ptr())
-        })
-        .map(Mthread)
-    }
-
-    pub fn id(&self) -> u64 {
-        unsafe { spdk_thread_get_id(self.0.as_ptr()) }
-    }
-    ///
-    /// # Note
-    ///
-    /// Avoid any blocking calls as it will block the whole reactor. Also, avoid
-    /// long-running functions. In general if you follow the nodejs event loop
-    /// model, you should be good.
-    pub fn with<T, F: FnOnce() -> T>(self, f: F) -> T {
-        let th = Self::current();
-        self.enter();
-        let out = f();
-        if let Some(t) = th {
-            t.enter();
-        }
-        out
-    }
-
-    #[inline]
-    pub fn poll(&self) {
-        let _ = unsafe { spdk_thread_poll(self.0.as_ptr(), 0, 0) };
-    }
-
-    #[inline]
-    pub fn enter(&self) {
-        unsafe { spdk_set_thread(self.0.as_ptr()) };
-    }
-
-    #[inline]
-    pub fn exit(&self) {
-        unsafe { spdk_set_thread(std::ptr::null_mut()) };
-    }
-
-    pub fn current() -> Option<Mthread> {
-        NonNull::new(unsafe { spdk_get_thread() }).map(Mthread)
-    }
-
-    pub fn name(&self) -> &str {
-        unsafe {
-            std::ffi::CStr::from_ptr(spdk_thread_get_name(self.0.as_ptr()))
-                .to_str()
-                .unwrap()
-        }
-    }
-
-    pub fn into_raw(self) -> *mut spdk_thread {
-        self.0.as_ptr()
-    }
-
-    /// destroy the given thread waiting for it to become ready to destroy
-    pub fn destroy(self) {
-        debug!("destroying thread {}...{:p}", self.name(), self.0);
-        unsafe {
-            spdk_set_thread(self.0.as_ptr());
-            // set that we *want* to exit, but we have not exited yet
-            spdk_thread_exit(self.0.as_ptr());
-
-            // now wait until the thread is actually exited the internal
-            // state is updated by spdk_thread_poll()
-            while !spdk_thread_is_exited(self.0.as_ptr()) {
-                spdk_thread_poll(self.0.as_ptr(), 0, 0);
-            }
```

Comment:
Ah so that changes on the other repo came from here...
My understanding is very limited, but is this safe to poll in a tight loop without letting the reactor get a chance to run too?

Reply:
Sorry Tiago, I cannot answer this question :-) I just moved code around to make structure more clear. Right now, I cannot reason how good any particular part of our reactor/thread code it. I remember Jeffry telling me that some parts are actually not quite good, and should probably be reworked. He wanted Zeeshan to do that, but unfortunately he didn't finish his work.

Perhaps, we should add a ticket to our Tech Debt epic to make investigation in that area. As for EB2 release, I'd rather keep as is.